contrib/gorilla/mux/mux.go

```diff
@@ -116,7 +116,7 @@ func (r *Router) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		spanopts = append(spanopts, headerTagsFromRequest(req))
 	}
 	resource := r.config.resourceNamer(r, req)
-	httputil.TraceAndServe(r.Router, &httputil.TraceConfig{
+	httptrace.TraceAndServe(r.Router, &httptrace.ServeConfig{
```

Comment:
I wonder if we can improve this, now that it has become public API. For example: users might leave essential config blank inside `ServeConfig`, which will cause this function to fail, or even panic.

Do we now need to do some sort of validation? And perhaps return an error that the user should check? For example, some fields are mandatory in `ServeConfig` (most of them are).

Also, we need to update and improve the documentation for it. For both this function (`TraceAndServe`) and for the config struct (`ServeConfig`). We need to clearly explain which field is mandatory, what it does, and which one is optional by using the proper documentation style (starting with the identifier, etc). Now, it's just some inline text to the right of each field.

Reply:
@gbbr Agree. We need clear document of these function and struct. 
But I wonder returning error from `TranceAndServe`. It is used for tracing request/response process. So I think we should not prevent customer main purpose that is  receive request and return response. 

So how about it ?

```go
func TraceAndServe(h http.Handler,  w http.ResponseWriter, req *http.Request, opts ...ServeOpt) {
}
```

- `Request` and  `ResponseWriter` are mandatory. So We need to pass as arguments immediately.
- `Service`, `Resource`, `QueryParams`, `FinishOpts` and `SpanOpts` are not mandatory. Especially if `Service` and `Resource` are empty, it might be happen lack of some tags. But it is not critical I think.
- This project use option function patterns, so we can also use it here.

Reply:
I agree it's nicer to not have any error return. I'd say that for the sake of consistency, it's probably best to go with this one:
```go
func TraceAndServe(h http.Handler,  w http.ResponseWriter, req *http.Request, opts ...ServeOption)
```
I think it makes sense to add `ServeOption` functional options since we don't use structs for config in dd-trace-go. While I do think structs would be better here, I think it might be best to stay consistent.

Reply:
Actually, hold on. Adding another type for options already creates problems with clashes. See for example `WithService`. It can end up confusing. So perhaps still, the struct would be better 🤔

Reply:
Yes, we can declare options like that. But I agree it is not clear for user that exist "WithService" and "WithServeService". But it is observed as type, so user cant pass wrong option.

My another concern of functional options is performance. This config is evaluated every request/response process. So in this mean, just struct is good.

```go
type serveConfig struct {
	service     string                    // service name
	resource    string                    // resource name
	queryParams bool                      // specifies that request query parameters should be appended to http.url tag
	finishOpts  []ddtrace.FinishOption    // span finish options to be applied
	spanOpts    []ddtrace.StartSpanOption // additional span options to be applied
}

type ServeOption func(*serveConfig)

func WithServeService(name string) ServeOption {
	return func(cfg *serveConfig) {
		cfg.service = name
	}
}
```

Reply:
> My another concern of functional options is performance. This config is evaluated every request/response process. Should we keep simple as struct ?

Yes. Very good concern. On the spot 🙂 I think it's clear what we need to do then!

Reply:
OK, I'll change to like that. Is it right ?

```go
func TraceAndServe(h http.Handler,  w http.ResponseWriter, req *http.Request, cfg *ServeConfig) {
}

type ServeConfig struct {
	Service     string                    // service name
	Resource    string                    // resource name
	QueryParams bool                      // specifies that request query parameters should be appended to http.url tag
	FinishOpts  []ddtrace.FinishOption    // span finish options to be applied
	SpanOpts    []ddtrace.StartSpanOption // additional span options to be applied
}
```